### ISSUE TITLE
Panic while generating proofs for testnet

### ISSUE THREAD

**Issue by dergoegge** [mit-dci/utreexo#130]
```
On block : 190001
On block : 200001
remap forest 19 rows -> 20 rows
remap forest 20 rows -> 21 rows
panic: genDels block 206482 87 txs but 0 rev txs

goroutine 1 [running]:
main.main()
        /Users/niklas/go/src/github.com/mit-dci/utreexo/cmd/main.go:71 +0x525
```
I am getting this error/panic message on current master(https://github.com/mit-dci/utreexo/commit/cb5c021b3d001a5f5b4dd3eedce8aac034a5dbd2)  when running the `cmd` binary on testnet.

Might be related to #114 

Might also be caused by:
* *running on macOS*
* *syncing to testnet using the bitcoin core master branch(https://github.com/bitcoin/bitcoin/commit/dc5333d31f280e09bb1e8cdacfbe842f4ab9e69b), which is not guaranteed to be completely stable.*

**Comment by kcalvinalvin** [comment]
Ah this is because of block order misalignment in ```blk*.dat``` files and ```rev*.dat``` files. Currently the master branch of Utreexo uses the ```rev*.dat``` files for fetching tx validation data for blocks but there's isn't *anything* in the ```rev*.dat```` file to indicate the height of the rev block.

So by the error message you've reproduced, it seems like the ```rev*.dat``` file is off order at height 206482. This is normal behavior that the current Utreexo code isn't handling right now.

We're aware and talked about the issue on the weekly meetings and going forward, instead of relying on ```rev*.dat``` files, we'll replace them with rpc calls to bitcoin core.

**Comment by adiabat** [comment]
Yeah this is most likely the same error as #114 ... it's a tricky one to fix.
The rev.dat files are mostly in the same order as the blk.dat files, but in some cases it's not.  The tricky part is that just given the rev and blk files, there's no sure way to figure out the ordering of the rev data.  The blk files are at least a blockchain, where hashes point to the previous block, so even when they're out of order you can figure out the correct order from the hashes.  The rev blocks don't have hashes that mean anything though.

One way to fix this that will work like 99+% of the time would be to check that the block and rev block at the assumed same height match in terms of transactions and inputs, and if they don't keep the rev block in ram in a map until we can assign it.

Something like: check block, make an int slice x where len(x) == len(blk.Transactions) and x[i] == len(blk.Transactions[i].TxIn).  Generate a similar slice y for the rev block, and check that the slices are the same.  If they're not, cache the rev and reorder it when a matching y comes up (hopefully the next rev block)

This isn't foolproof though.  It's *possible* that two blocks have been transposed with the exact same number of transactions and the same number of inputs in each, which this won't detect.  However, the current code won't detect this bug either!  Or at least we'd get a different error, that there was an invalid proof.  And we haven't seen that.  Also it seems super unlikely -- the rev / blk mismatch seems to happen like once out of 500K+ blocks (I haven't gotten it on my blocks folder for testnet or mainnet) and also having that coincide with blocks with identical input fingerprints seems pretty unlikely.

However that's not 100% so we do need a better long term solution (probably have to dig in to bitcoin-core's levelDB indexes) but could be a good enough fix for now.

**Comment by ecdsa** [comment]
Same issue here, on mainnet:
```
On block : 110001
remap forest 17 rows -> 18 rows
leveldb: closed
panic: genDels block 119124 18 txs but 6 rev txs

goroutine 1 [running]:
main.main()
	/root/go/src/github.com/mit-dci/utreexo/cmd/main.go:71 +0x530
```

**Comment by adiabat** [comment]
Hopefully fixed in PR #144, please check if it works now!
